File: src/boil/detectors/make_no_rule.go
// Detector for make "No rule to make target" errors.
package detectors

// MakeNoRuleDetector detects make "No rule to make target" errors
type MakeNoRuleDetector struct {
	*BaseDetector
}

func NewMakeNoRuleDetector() (*MakeNoRuleDetector, error) {
	base, err := NewBaseDetector(
		"MakeNoRuleDetector",
		100,
		map[string]string{
			"make_no_rule": `make(?:\[\d+\])?: \*\*\* No rule to make target '(?P<target>[^']+)'\.\s+Stop\.`,
		},
		[]DetectorExample{
			{
				Name:     "make_no_rule",
				Input:    "make: *** No rule to make target 'test'.  Stop.",
				ClueType: "make_no_rule",
				Context:  map[string]string{"target": "test"},
			},
			{
				Name:     "make_no_rule_with_level",
				Input:    "make[2]: *** No rule to make target 'build'.  Stop.",
				ClueType: "make_no_rule",
				Context:  map[string]string{"target": "build"},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &MakeNoRuleDetector{base}, nil
}

File: src/pipeline/detectors/make_no_rule.py
"""
Detector for make no rule errors.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class MakeNoRuleDetector(Detector):
    """
    Detect make errors when no rule exists for a target (often means Makefile is missing).

    Matches patterns like:
    - make: *** No rule to make target 'test'.  Stop.
    """

    PATTERNS = {
        "make_no_rule": r"make(?:\[\d+\])?: \*\*\* No rule to make target '(?P<target>[^']+)'\.\s+Stop\.",
    }

    EXAMPLES = [
        (
            "make: *** No rule to make target 'test'.  Stop.",
            {
                "clue_type": "make_no_rule",
                "confidence": 1.0,
                "context": {
                    "target": "test",
                },
            },
        ),
    ]
